meson.build

```diff
@@ -0,0 +1,252 @@
+  'src/google/protobuf/compiler/java/java_primitive_field_lite.cc',
+  'src/google/protobuf/compiler/java/java_service.cc',
+  'src/google/protobuf/compiler/java/java_shared_code_generator.cc',
+  'src/google/protobuf/compiler/java/java_string_field.cc',
+  'src/google/protobuf/compiler/java/java_string_field_lite.cc',
+  'src/google/protobuf/compiler/js/js_generator.cc',
+  'src/google/protobuf/compiler/js/well_known_types_embed.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_enum.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_enum_field.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_extension.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_field.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_file.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_generator.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_helpers.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_map_field.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_message.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_message_field.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_oneof.cc',
+  'src/google/protobuf/compiler/objectivec/objectivec_primitive_field.cc',
+  'src/google/protobuf/compiler/php/php_generator.cc',
+  'src/google/protobuf/compiler/plugin.cc',
+  'src/google/protobuf/compiler/plugin.pb.cc',
+  'src/google/protobuf/compiler/python/python_generator.cc',
+  'src/google/protobuf/compiler/ruby/ruby_generator.cc',
+  'src/google/protobuf/compiler/subprocess.cc',
+  'src/google/protobuf/compiler/zip_writer.cc',
+]
+libprotoc = library(
+  'protoc',
+  libprotoc_src,
+  cpp_args: flags,
+  link_with: [libprotobuf_lite, libprotobuf],
+  include_directories: include_directories('src')
+)
+protoc_dep = declare_dependency(
+  link_with: [libprotobuf_lite, libprotobuf, libprotoc],
+  include_directories: include_directories('src')
+)
+
+
+protoc_src = [ 'src/google/protobuf/compiler/main.cc' ]
+protoc = executable(
+  'protoc',
+  protoc_src,
+  cpp_args: flags,
+  include_directories: include_directories('src'),
+  link_with: [libprotobuf_lite, libprotoc]
+)
+
+
+if meson.is_cross_build()
+  protoc_native = executable(
+    'protoc-native',
+    [libprotobuf_lite_src, libprotobuf_src, libprotoc_src, protoc_src],
+    native: true,
+    cpp_args: flags,
+    include_directories: include_directories('src'),
+    dependencies: dependency('threads', native: true),
+  )
+  protoc_gen = generator(protoc_native,
+    output: ['@BASENAME@.pb.cc', '@BASENAME@.pb.h'],
+    arguments: ['@EXTRA_ARGS@', '--proto_path=' + meson.current_source_dir()/'src', '--cpp_out=@BUILD_DIR@', '@INPUT@']
+  )
+else
+  protoc_gen = generator(protoc,
+    output: ['@BASENAME@.pb.cc', '@BASENAME@.pb.h'],
+    arguments: ['@EXTRA_ARGS@', '--proto_path=' + meson.current_source_dir()/'src', '--cpp_out=@BUILD_DIR@', '@INPUT@']
+  )
+endif
```

Comment:
Better to set the protoc_native variable to the regular executable, if it's a non-cross build.

Then define one generator outside of the if.